src/index/txindex.cpp

```diff
@@ -0,0 +1,252 @@
+    }
+
+    if (!chain_tip) {
+        m_synced = true;
+        return true;
+    }
+
+    uint256 best_block_hash;
+    if (!m_db->ReadBestBlockHash(best_block_hash)) {
+        FatalError("%s: Failed to read from tx index database", __func__);
+        return false;
+    }
+
+    if (best_block_hash.IsNull()) {
+        return true;
+    }
+
+    auto it = mapBlockIndex.find(best_block_hash);
+    if (it == mapBlockIndex.end()) {
+        FatalError("%s: Last block synced by txindex is unknown", __func__);
+        return false;
+    }
+
+    const auto pindex = it->second;
+    m_best_block_index = pindex;
+    if (pindex->GetAncestor(chain_tip->nHeight) == chain_tip) {
+        m_synced = true;
+    }
+
+    return true;
+}
+
+static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev)
+{
+    AssertLockHeld(cs_main);
+
+    if (!pindex_prev) {
+        return chainActive.Genesis();
+    }
+
+    auto pindex = chainActive.Next(pindex_prev);
+    if (pindex) {
+        return pindex;
+    }
+
+    return chainActive.Next(chainActive.FindFork(pindex_prev));
+}
+
+void TxIndex::ThreadSync()
+{
+    auto pindex = m_best_block_index.load();
+    if (!m_synced) {
+        auto& consensus_params = Params().GetConsensus();
+
+        int64_t last_log_time = 0;
+        while (true) {
+            if (m_interrupt) {
+                return;
+            }
+
+            {
+                LOCK(cs_main);
+                auto pindex_next = NextSyncBlock(pindex);
+                if (!pindex_next) {
+                    m_best_block_index = pindex;
+                    m_synced = true;
+                    break;
+                }
+                pindex = pindex_next;
+            }
+
+            int64_t current_time = GetTime();
+            if (last_log_time + SYNC_LOG_INTERVAL < current_time) {
+                LogPrintf("Syncing txindex with block chain from height %d\n", pindex->nHeight);
+                last_log_time = current_time;
+            }
+
+            CBlock block;
+            if (!ReadBlockFromDisk(block, pindex, consensus_params)) {
+                FatalError("%s: Failed to read block %s from disk",
+                           __func__, pindex->GetBlockHash().ToString());
+                return;
+            }
+            if (!WriteBlock(block, pindex)) {
+                FatalError("%s: Failed to write block %s to tx index database",
+                           __func__, pindex->GetBlockHash().ToString());
+                return;
+            }
+        }
+    }
+
+    if (pindex) {
+        LogPrintf("txindex is enabled at height %d\n", pindex->nHeight);
+    } else {
+        LogPrintf("txindex is enabled\n", pindex->nHeight);
+    }
+}
+
+bool TxIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint256, CDiskTxPos>> vPos;
+    vPos.reserve(block.vtx.size());
```

Comment:
nit: move this to the ctor (it can take a size to reserve)

Reply:
I'm pretty sure the `std::vector` constructor that takes a `size_t` creates the vector with `n` elements rather than having a size of 0 and a capacity of `n`.

Reply:
That's not strictly accurate:

```
$ g++ --version
g++ (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ cat reserve.cc 
#include <iostream>
#include <vector>

struct Foo {
    int x;
    Foo() :x(1) {}
};

int main(int argc, char **argv) {
    std::vector<Foo> foovec;
    foovec.reserve(10);
    std::cout << foovec[5].x << std::endl;
    std::cout << foovec.at(5).x << std::endl;
    return 0;
}

$ g++ reserve.cc 

$ ./a.out 
0
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 5) >= this->size() (which is 0)
Aborted (core dumped)
```

In this test `Foo` has a constructor that sets `x` to 1, but as you can see from the first line of output it prints 0 implying that the ctor wasn't actually called. The `.at()` check fails because the element is out of bounds compared to the true size. The standard calls this ["default insertable"](http://en.cppreference.com/w/cpp/concept/DefaultInsertable) which I think just means "not allocator aware" in which case the default allocator is allowed to just create space for the item.

Either way it's a nit that you can ignore, but I think the suggestion I made is safe and slightly faster, assuming you actually initialize the elements before accessing them.

Reply:
Have you tried that code sample with `std::vector<Foo> foovec(10);` instead of `std::vector<Foo> foovec; foovec.reserve(10);`? The behavior is different. I assume that is the suggestion you are making, though maybe I misunderstood.

Reply:
Have you tried that code sample with std::vector<Foo> foovec(10); instead of std::vector<Foo> foovec; foovec.reserve(10);? The behavior is different. I assume that is the suggestion you are making, though maybe I misunderstood.